### DISCUSSION TITLE
Supporting AI tools

### DISCUSSION THREAD

**Discussion by antonpirker** [getsentry/sentry-python#3007] [Q&A]
Hey everybody!

Not long ago we added an [OpenAI integration](https://docs.sentry.io/platforms/python/integrations/openai/) to the Sentry SDK. Right now we have PRs for two more AI related integrations open, [Anthropic](https://github.com/getsentry/sentry-python/pull/2831) and [Langchain](https://github.com/getsentry/sentry-python/pull/2911).

If you work in the AI sector or are an AI enthusiast, I would like to ask you a couple of questions:
- What data from your AI project would you like to see in Sentry?
- What data would help to improve your AI pipelines, the training of your models, help you improve your AI project?
- What other AI related features would you love to see in Sentry?
- Are there any other tools you would like to see automatic instrumentation for?

Let's get a discussion going, so together we can build kick ass AI support into Sentry!

**Comment by smeubank** [comment]
After talking with some folks internally and from the community (downloads and stars etc), I believe the most useful platforms for folks would be as follows, grouped by product/service, and then popularity within. ✅  Means we already have a PR in the works for them

### LLM Providers
- openai ✅ 
- huggingface-hub https://github.com/huggingface/huggingface_hub
- anthropic ✅ 

### Framework
- langchain ✅ 
- llama-index https://github.com/run-llama/llama_index
- haystack https://github.com/deepset-ai/haystack

### Vector DBs
- chromadb https://github.com/chroma-core/chroma
- cohere https://github.com/cohere-ai
- qdrant-client https://github.com/qdrant/qdrant-client
- weaviate-client https://github.com/weaviate/weaviate-python-client

**Reply by smeubank** [reply]
@czyber fyi for the contribution on the Anthropic integration

**Reply by smeubank** [reply]
further what I think is a very interesting development in AI obersvability ecosystem is [OpenLLMetry](https://github.com/traceloop/openllmetry?tab=readme-ov-file#llm-providers), which is built on top of OpenTelemetry 🔭 

I would need to investigate more, but generally open standards standards for instrumentation, APIs, and span schemas can be a positive for developers looking to add observability to their stacks. 🎉 

Some thoughts I have -> 
- can we learn from, or even use portions of these libraries as dependencies of our own SDKs?
  - could we contribute upstream then?
- I would like to see it more officially part of OpenTelemetry (CNCF)
- Could we work with all AI related library maintainers to emit OpenTelemetry traces natively (provide hooks), as a similar pattern to what we are seeing from many other frameworks and libaries ( Next.js for [example](https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry) )

**Reply by smeubank** [reply]
@kamilogorek i wonder if you know anyone at Supabase that might have some thoughts on telemetry for their Vector DBs? I see from the docs Supabase suggests to use https://github.com/pgvector/pgvector. Any other libraries that would be interesting? Could supabase provide traces via some hook we could latch onto?

**Reply by czyber** [reply]
For Vector DBs, i also know about [Deeplake](https://github.com/activeloopai/deeplake)

**Reply by colin-sentry** [reply]
note: django-haystack is not actually AI related, it's another project with the same name

I'll be doing cohere & huggingface integrations later this week

**Reply by smeubank** [reply]
@czyber so if you wanted to work on another integration, since @colin-sentry will do hugging face and cohere

This also means there will be a model for how we'd like to see instrumentation for a Vector DB. Let us know what you're interested in, and we'll try to support!

**Reply by czyber** [reply]
Hey hey! Sure I'll give a vector db a spin. If you have any preferences on which to integrate first, I'm open to do any of them. Otherwise I'll have time on the weekend to have a deeper look :+1:

**Reply by antonpirker** [reply]
@colin-sentry you probably know best what vector db @czyber could look into. You do cohere. Chroma is the one that is included in Langchain, so maybe this is a good candidate? Or should we wait until you do a cohere integration and just copy it and modify it to fit another db?

**Reply by colin-sentry** [reply]
I was just going to do the cohere LLM, not their vector DB yet.

We do have the embeddings measured in the OpenAI integration, which is most of the interesting bits for most vector DBs

**Reply by smeubank** [reply]
in the future we may have better/more native OTel support in the Python SDK

but until then :) @czyber if you want it might be interesting to work on chromadb, and refer the implementation here for inspiration

https://github.com/traceloop/openllmetry/tree/main/packages/opentelemetry-instrumentation-chromadb

**Reply by czyber** [reply]
Sure thing!

**Comment by smeubank** [comment]
### Interesting use cases:

the following I am regurgitating from @czyber 😆  for full transparency 🚀 

> An interesting metric to observe for LLMs in general could be the output tokens from the LLM categorized by prompts. 
>
> Lets take the following example: 
I have the use case that I want the LLM to create studying material for students (e.g. task descriptions for math problems), 
>
> The prompt is a template, and has some dynamic inputs (such as 'output_language', 'grade', 'topic'). It would be interesting to have 
some metrics (maybe even alerts) if the LLM responses deviate from the expectation value of the output_token count.

**Comment by smeubank** [comment]
Hugging Face does provide some support in the way that you can indicate an OTLP endpoint ([Text Generation Interface: Distributed Tracing](https://github.com/huggingface/text-generation-inference?tab=readme-ov-file#distributed-tracing)). Which appear to be what openLLMetry takes advantage of [here](https://github.com/traceloop/openllmetry/tree/main/packages/opentelemetry-instrumentation-transformers)

**Comment by antonpirker** [comment]
Just a quick update:
- Our LLM monitoring is now out of Beta: https://docs.sentry.io/product/insights/llm-monitoring/
- The Python libraries we support right now can be found here: https://docs.sentry.io/platforms/python/integrations/#ai

**Comment by iloveitaly** [comment]
@antonpirker the documentation isn't clear: If I have the py openai library installed, will monitoring automatically be setup or do I need to `@ai_track("My AI pipeline")` each function which makes an openai call?

**Reply by nsaef** [reply]
The same goes for LangChain: the way the documentation is written, I would assume it to track my LLM calls automatically. However, the "LLM Monitoring" section only showed up once I explicitely initialized Sentry with the LangChain integration, and it's not creating any traces without using the decorators.

Relevant excerpt from the docs:
> The Sentry LLM Monitoring feature relies on the fact that you have an orchestrator (like LangChain) creating pipelines of one or more LLMs (such as gpt-4). [...]
> 
> If you're using a provider like OpenAI without an orchestrator like LangChain, you'll need to manually create pipelines with the @ai_track annotation.

**Reply by smeubank** [reply]
hi @iloveitaly and @nsaef 

thanks for this feedback. We discussed this in planning today, I don't have an ETA, but we did create an issue. Generally i think we should better communicate what auto-instrumentation does from the SDK, and what we expect form users. Feel free to comment any corrections or feedback also on that issue.

https://github.com/getsentry/sentry-docs/issues/13167

**Comment by invke** [comment]
Hi @antonpirker, we are typically using https://github.com/griptape-ai/griptape as our AI framework. Is there plans to support integration there? The current project I'm working on is using OpenAI models (for now) via Griptape, which in turn uses the OpenAI SDK, so I had thought perhaps it would work. However, it doesn't seem to report token usage, has non realistic 0.02ms durations and of course then no cost estimates. I figure perhaps it's just an incompatibility with how Griptape operates around the SDKs.

Did think about them both supporting Open Telemetry but figured I'd check here first as I wasn't immediately sure how that'd work and was time-bound.

**Reply by smeubank** [reply]
hi @invke 

It isn't planned. We are working on OpenTelemetry support, but OpenAI is not necessarily a core feature support of this work, https://github.com/getsentry/projects/issues/68. I don't know that OTel has native support for AI libraries today. There is [OpenLLMetry](https://github.com/traceloop/openllmetry) which is not part of OTel, and is 3rd party. 

i also created an issue to track this topic, and let the team consider it, and any more context you might want to add to the issue.

https://github.com/getsentry/sentry-python/issues/4330

**Comment by Luke31** [comment]
Hi @antonpirker . I am not sure if this thread is still being followed, but some feedback from my side. Sorry for the long post.

We are using OpenAI with Langchain. After following the documentation I found it a bit difficult to start using so I created a PR to update the documentation: https://github.com/getsentry/sentry-docs/pull/13905

Currently two use-cases aren't easily covered by current functionality:

### Used tokens by tag

See number of used tokens by custom tag or user-id. Currently we can only see the total used token by project and enviornment. There is no method to further drill down in the [AI LLM-Monitoring tab in sentry](https://sentry.io/insights/ai/llm-monitoring/). 

<img width="1461" alt="image" src="https://github.com/user-attachments/assets/0620534c-c613-4d40-ad82-9c2393c312cd" />

It would be nice to have further filters similar to traces tab: 

<img width="618" alt="image" src="https://github.com/user-attachments/assets/326d675a-e4cd-4f5c-ac34-43fcbe20017f" />

### Quick overview to view multiple input and output prompts

Currently input and output prompts are hard to find from the AI pipeline tab. Multiple clicks and identifying the correct span in the traces view to find the prompt is cumbersome. It is maybe enough for finding a bug but insufficient for frequent use, especially if the prompt is long. The UI is not made to display long prompts. 

![image](https://github.com/user-attachments/assets/0b18bf5a-05dd-4dfc-8345-97f6ebd27251)

If the prompt is too long, it is also cut off (maybe configureable?):

<img width="553" alt="image" src="https://github.com/user-attachments/assets/200162a2-7bf8-4c4d-b253-4ca993626d86" />

When developing an AI pipeline, I'm mainly interested in the data in-and output and prompt in-and output.

Of course the requirements depend on each developer and maybe sentry is not the correct tool for these use-cases.
Thank you!

**Reply by smeubank** [reply]
hey @Luke31 

thanks for this! I will make sure someone gets eyes on this

**Reply by antonpirker** [reply]
Hey @Luke31 we are working on a better AI experience at the moment, so stay tuned, it will get better. 

For the "tokens used" by tag or user, I am not sure but I asked our backend folks. I will keep you posted!
